Fix duplicate css when enable critical path

mrtuvn opened the pull request:
<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This PR continue great works https://github.com/magento/magento2/pull/27211 by @krzksz in latest codebase
This PR aim to bring patch fix issue css show twice. Before the fix magento place redundant link style at bottom of page before body close end

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#26498

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Enable critical path css
2. Run deploy static content + clean cache
3. Go to homepage inspect in browser element "page-wrapper"
4. We will see 2 file styles-l.css have style apply for that element
### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

mrtuvn commented:
@magento run all tests

ihor-sviziev commented:
@magento run all test

@mrtuvn does this PR replace https://github.com/magento/magento2/pull/27211 or what is the purpose of it?

ihor-sviziev commented:
@magento run all tests

mrtuvn commented:
@ihor-sviziev just continue works by @krzksz after long inactive

mrtuvn commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php:
```diff
@@ -32,48 +35,94 @@ public function __construct(ScopeConfigInterface $scopeConfig)
+        $content = (string)$httpResponse->getContent();
+        $headCloseTag = '</head>';
+
+        $headEndTagFound = strrpos($content, $headCloseTag) !== false;
```
From my experience `<body>` tag usually is much bigger than `<head>`, so I suggest to use `strpos` function instead
```suggestion
        $headEndTagFound = strpos($content, $headCloseTag) !== false;
```

ihor-sviziev commented on app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php:
```diff
@@ -32,48 +35,94 @@ public function __construct(ScopeConfigInterface $scopeConfig)
+     *
+     * @param string $content
+     */
+    private function extractLinkTags(&$content): string
```
```suggestion
    private function extractLinkTags(string &$content): string
```

ihor-sviziev commented on app/code/Magento/Theme/Test/Unit/Controller/Result/AsyncCssPluginTest.php:
```diff
@@ -98,48 +137,61 @@ public function sendResponseDataProvider(): array
+     * @dataProvider renderResultDataProvider
      */
-    public function testBeforeSendResponse($content, $isSetFlag, $result): void
+    public function testAfterRenderResult($content, $isSetFlag, $result): void
```
```suggestion
    public function testAfterRenderResult(string $content, bool $isSetFlag, string $result): void
```

ihor-sviziev commented on app/code/Magento/Theme/view/frontend/templates/js/css_rel_preload.phtml:
```diff
@@ -2,7 +2,10 @@
 /**
  * Copyright © Magento, Inc. All rights reserved.
  * See COPYING.txt for license details.
+ * @deprecated since 100.4.0
```
```suggestion
 * @deprecated as polyfill isn't used anymore
 * @since 100.4.0
```

ihor-sviziev reviewed (changes requested):
Hi @mrtuvn,
In general your changes looks good, but I have few suggestions, please review them

ihor-sviziev commented:
Could you also squash your changes into one or few commits in order not to have 13 merge commits?

mrtuvn commented:
i have squashed commits and repick all to one commit. Ready for review and test

ihor-sviziev commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7678](https://jira.corp.magento.com/browse/ENGCOM-7678) has been created to process this Pull Request

ihor-sviziev commented:
@magento run all tests

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Enable critical path CSS ( Admin -> Stores -> Configuration -> Advanced -> Developer -> Css Settings, set **Use CSS critical path** to **Yes**);
2. Run deploy static content + clean cache;
3. Go to homepage inspect in browser element "page-wrapper"

**Before:** :heavy_multiplication_x: `styles-m.css` and `styles-l.css` styles show up multiple times in developer tools.

![Screenshot from 2020-06-16 10-46-22](https://user-images.githubusercontent.com/51679138/84757572-fa8daf00-afcc-11ea-97b4-2cf601bb4085.png)

**After:** :heavy_check_mark: The stylesheets are only loaded once

![Screenshot from 2020-06-16 11-34-57](https://user-images.githubusercontent.com/51679138/84757654-11cc9c80-afcd-11ea-8c3e-6baf7938b378.png)

:heavy_check_mark: all external stylesheets are located in the same order at the end of <head>, after inline critical CSS

![screenshot_64](https://user-images.githubusercontent.com/51679138/84758002-8acbf400-afcd-11ea-9cf7-294170119f14.png)

Also was tested on the Blank theme and with sample data.

engcom-Charlie commented:
It needs to approve.
**Semantic Version Checker** failed because of `css_rel_preload_script` block was removed.

mrtuvn commented:
@DrewML @omiroshnichenko Can you guys take a look on this Thanks

mrtuvn commented:
@sshymko @lbajsarowicz what do you guys think can we need improve in code ?

lbajsarowicz commented on app/code/Magento/Theme/view/frontend/templates/js/css_rel_preload.phtml:
```diff
@@ -2,7 +2,11 @@
  * Copyright © Magento, Inc. All rights reserved.
  * See COPYING.txt for license details.
+ * @deprecated as polyfill isn't used anymore
+ * @since 100.4.0
```
As far as I remember - the automated tool adds the Version number to `@deprecated`. It's enough if you add `@deprecated` then.

lbajsarowicz commented on app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php:
```diff
@@ -32,48 +35,94 @@ public function __construct(ScopeConfigInterface $scopeConfig)
+            $style = sprintf(
+                '<link rel="stylesheet" media="print"' .
+                ' onload="this.onload=null;this.media=\'%s\'"' .
+                ' href="%s">',
```
Can't this be squashed into single line?

lbajsarowicz commented on app/code/Magento/Theme/Controller/Result/AsyncCssPlugin.php:
```diff
@@ -32,48 +35,94 @@ public function __construct(ScopeConfigInterface $scopeConfig)
+            $content = str_replace($style, '', $content);
+
+            preg_match('@href=("|\')(.*?)\1@', $style, $hrefAttribute);
+            $href = $hrefAttribute[2];
```
I'm not sure about this code behaviour when there's invalid syntax provided.
Would you mind adding extra condition:
```suggestion
            if (!preg_match('@href=("|\')(.*?)\1@', $style, $hrefAttribute)) {
                throw new \Exception('Invalid <style> syntax provided');
            }
            $href = $hrefAttribute[2];
```

or something similar.

lbajsarowicz reviewed (changes requested):
Just minor adjustments are necessary.

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

lbajsarowicz reviewed (dismissed):
:heavy_check_mark: Thank you for your contribution!

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7678](https://jira.corp.magento.com/browse/ENGCOM-7678) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

engcom-Charlie commented:
It needs to approve.
Semantic Version Checker failed because of css_rel_preload_script block was removed.

mrtuvn commented:
@magento run all tests

mrtuvn commented:
hi any update on this

mrtuvn commented:
@magento run Integration Tests

lbajsarowicz reviewed (dismissed):
I find the errors reported by tests as unrelated.
:heavy_check_mark: Thank you for your contribution!

magento-engcom-team commented:
Hi @lbajsarowicz, thank you for the review.
[ENGCOM-7678](https://jira.corp.magento.com/browse/ENGCOM-7678) has been created to process this Pull Request

ihor-sviziev commented:
@mrtuvn @lbajsarowicz this PR still waiting approval for SVC failure

lbajsarowicz commented:
@ihor-sviziev I know, but from my side the CR is actually :heavy_check_mark: green. Now it's up to Magento Commeng to approve that :)

mrtuvn commented:
@ihor-sviziev @lbajsarowicz seem we have fail on intergration tests
![Screenshot from 2020-07-30 17-00-57](https://user-images.githubusercontent.com/1908873/88910214-5747df00-d286-11ea-867a-74faa7231a3b.png)

mrtuvn commented:
@magento run all tests

mrtuvn commented:
@magento run all tests

ihor-sviziev commented on app/code/Magento/Theme/view/frontend/layout/default_head_blocks.xml:
```diff
@@ -18,7 +18,6 @@
                     <argument name="criticalCssViewModel" xsi:type="object">Magento\Theme\Block\Html\Header\CriticalCss</argument>
                 </arguments>
             </block>
-            <block name="css_rel_preload_script" ifconfig="dev/css/use_css_critical_path" template="Magento_Theme::js/css_rel_preload.phtml"/>
```
Unfortunately this is backward incompatible change that leads to SVC failure which is not allowed in patch releases. Could you revert block removal and just put the empty view file instead with only deprecation notice, so it will not output anything?

And in future this block could be removed in next 2.5 or 2.6 release

mrtuvn replied:
i can remove this block in layout xml so we don't need to empty template
Could add some note in layout for later refactor

ihor-sviziev replied:
good idea, but block with name `css_rel_preload_script` should be there.

mrtuvn replied:
yes i have updated in latest commit

ihor-sviziev replied:
I think even adding remove statement could break customizations, I suggest to go with one of those solutions:
- update view to not output anything
- remove “template” from the block

mrtuvn replied:
ok i have removed it from template now deprecated template will not output anything

ihor-sviziev commented on app/code/Magento/Theme/view/frontend/layout/default.xml:
```diff
@@ -101,7 +101,7 @@
         <referenceContainer name="main">
             <container name="content.top" label="Main Content Top">
-                <block name="main_css_preloader" as="main_css_preloader" template="Magento_Theme::html/main_css_preloader.phtml" ifconfig="dev/css/use_css_critical_path"/>
+                <block class="Magento\Framework\View\Element\Template" name="main_css_preloader" as="main_css_preloader" template="Magento_Theme::html/main_css_preloader.phtml" ifconfig="dev/css/use_css_critical_path"/>
```
I believe we don’t need this change, as this is default block type

ihor-sviziev commented on app/code/Magento/Theme/view/frontend/templates/html/main_css_preloader.phtml:
```diff
@@ -4,12 +4,18 @@
+ */
+
+$escaper = isset($escaper) ? $escaper : $block;
 ?>
```
I believe we don’t need this check, as it always available in magento 2.4

ihor-sviziev commented on app/code/Magento/Theme/view/frontend/templates/js/css_rel_preload.phtml:
```diff
@@ -25,4 +26,3 @@
 
 script;
 ?>
```
I think we don’t need to close php tag in this file at all, also on line 8 too

mrtuvn commented:
@magento run all tests

ihor-sviziev commented:
@magento run all tests

lbajsarowicz commented:
@mrtuvn & @ihor-sviziev If only one test was failing, you don't need to re-run all of them. Only when you confirm that the only one that was failing now passes, you can run the rest, to make the PR green. It's generally worth to reduce the load for build servers of Magento, because this affects also the stability of the environment (unfortunately).

ihor-sviziev commented:
@lbajsarowicz there was only one test failing, so I fixed it and re-ran all tests

mrtuvn commented:
yes i'm totally agree with that. Run all tests can take lot of times but after change in commit magento will reset all stage builds. Hope we will have all green after run.

mrtuvn commented on dev/tests/integration/testsuite/Magento/Framework/Interception/PluginListGeneratorTest.php:
```diff
@@ -94,7 +94,6 @@ public function testPluginListConfigGeneration()
             1 => [
                 0 => 'genericHeaderPlugin',
                 1 => 'asyncCssLoad',
```
@ihor-sviziev seem after this change plugin name asyncCssLoad already changed to intercept class 
Magento\Framework\View\Result\Layout instead previously Magento\Framework\App\Response\Http. So we should remove item asyncCssLoad. Is that correct ?
Also we should update plugin list added by @lbajsarowicz in another commit

mrtuvn commented:
@magento run all tests

ihor-sviziev commented:
@mrtuvn static test still failing, I believe we could just ignore this failure for now.

mrtuvn commented:
@lbajsarowicz can you give me some suggest about last static fail in intergration php

lbajsarowicz commented:
@mrtuvn 
```php
@SuppressWarnings(PHPMD.CouplingBetweenObjects)
```

mrtuvn commented:
@ihor-sviziev @lbajsarowicz updated 
Thanks for your help

ihor-sviziev commented:
@magento run Static Tests

Commit:
Fix css shows twice when css critical path enabled

update

fix permissions

Fix static check
update

fix fail tests

update
add extra condition check


update


update


update


update


update


update


Fix integration test
update


Fix static test fail

ihor-sviziev commented:
@magento run Static Tests

mrtuvn commented:
@magento run Static Tests

mrtuvn commented:
@magento run Database Compare, Functional Tests CE, Functional Tests EE, Functional Tests B2B, Integration Tests, Magento Health Index, Sample Data Tests CE, Sample Data Tests EE, Sample Data Tests B2B, Unit Tests, WebAPI Tests

ihor-sviziev commented:
@magento run Semantic Version Checker, Database Compare

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-7678](https://jira.corp.magento.com/browse/ENGCOM-7678) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Enable critical path CSS ( Admin -> Stores -> Configuration -> Advanced -> Developer -> Css Settings, set **Use CSS critical path** to **Yes**);
2. Run deploy static content + clean cache;
3. Go to homepage inspect in browser element "page-wrapper"

**Before:** :heavy_multiplication_x: `styles-m.css` and `styles-l.css` styles show up multiple times in developer tools.

![2020-08-10_14-23](https://user-images.githubusercontent.com/51679138/89781785-aad6ea00-db1c-11ea-8cb6-c10cbf9d961a.png)

**After:** :heavy_check_mark: The stylesheets are only loaded once

![2020-08-10_14-34](https://user-images.githubusercontent.com/51679138/89781801-b1656180-db1c-11ea-9ee9-271a058f4768.png)

:heavy_check_mark: all external stylesheets are located in the same order at the end of <head>, after inline critical CSS

![screenshot_64](https://user-images.githubusercontent.com/51679138/84758002-8acbf400-afcd-11ea-9cf7-294170119f14.png)

Also was tested on the Blank theme and with sample data.

xanka commented:
Hi @mrtuvn Did you test on production mode?

mrtuvn commented:
What problem with production that you got?
